types/src/bounded_future_queue.rs

```diff
@@ -0,0 +1,256 @@
+    futures: &'a BoundedFuturesUnordered<T>,
+}
+
+/// An async-friendly FuturesUnordered of bounded size. In contrast to a bounded channel,
+/// the queue makes it possible to modify and remove entries in it. In contrast to a FuturesUnordered,
+/// the queue makes it possible to enforce a bound on the number of items in the queue.
+pub struct BoundedFuturesUnordered<T: Future> {
+    /// The maximum number of entries allowed in the queue
+    capacity: usize,
+    /// The actual items in the queue.
+    queue: FuturesUnordered<T>,
+    /// This semaphore has as many permits as there are empty spots in the
+    /// `queue`, i.e., `capacity - queue.len()` many permits
+    push_semaphore: Semaphore,
+}
+
+impl<T: Future> std::fmt::Debug for BoundedFuturesUnordered<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "BoundedFuturesUnordered[cap: {}, free: {}]",
+            self.capacity,
+            self.push_semaphore.available_permits(),
+        )
+    }
+}
+
+unsafe impl<T: Future> Sync for BoundedFuturesUnordered<T> {}
+unsafe impl<T: Future> Send for BoundedFuturesUnordered<T> {}
+
+// We expect to grow this facade over time
+impl<T: Future> BoundedFuturesUnordered<T> {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            capacity,
+            queue: FuturesUnordered::new(),
+            push_semaphore: Semaphore::new(capacity),
+        }
+    }
+
+    /// Push an item into the queue. If the queue is currently full this method
+    /// blocks until an item is available
+    pub async fn push(&self, item: T) {
+        let permit = self.push_semaphore.acquire().await.unwrap();
+        self.queue.push(item);
+        permit.forget();
+    }
+
+    pub fn push_with_permit(&self, item: T, _permit: SemaphorePermit<'_>) {
+        self.queue.push(item);
+    }
+
+    /// Waits for queue capacity. Once capacity to push one future is available, it is reserved for the caller.
+    ///
+    /// WARNING: the order of pushing to the queue is not guaranteed. It must be enforced by the caller.
+    pub async fn reserve(&self) -> Result<UnorderedPermit<'_, T>, AcquireError> {
+        let permit = self.push_semaphore.acquire().await?;
+        Ok(UnorderedPermit {
+            permit,
+            futures: self,
+        })
+    }
+
+    /// Report the available permits
+    pub fn available_permits(&self) -> usize {
+        self.push_semaphore.available_permits()
+    }
+}
+
+impl<U, V, T: Future<Output = Result<U, V>>> BoundedFuturesUnordered<T> {
+    /// Creates a future that attempts to resolve the next item in the stream.
+    /// If an error is encountered before the next item, the error is returned instead.
+    pub fn try_next(&mut self) -> impl Future<Output = Result<Option<U>, V>> + '_ {
+        self.queue.try_next().inspect_ok(|val| {
+            if val.is_some() {
+                self.push_semaphore.add_permits(1)
```

Comment:
Why do we want to add a permit only in case of successful response?

Reply:
Because this is a [`TryStreamExt::try_next`](https://docs.rs/futures/latest/futures/stream/trait.TryStreamExt.html#method.try_next), so the semantics of this returning an `Option::None` are the same as [`StreamExt::next`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.next), i.e. the stream is empty. In which case its available permits is already at capacity, and adding more permits would raise it above the capacity.

Reply:
In [53a2a28](https://github.com/MystenLabs/narwhal/pull/724/commits/53a2a28af5faa6e6e75475c1616010823afba137) I've edited tests to reflect what happens with the capacity when you do `try_next()` past the end of the stream. I checked those tests do not pass if you remove the corresponding `if val.is_some()` guards.